show_activity() functions overhaul

OciXCrom opened the pull request:
Since all of the `show_activity` functions have always been a huge problem for users who want to add colored chat messages to their plugins, especially the default AMXX plugins, I decided to make a complete overhaul of the functions and make changes that will alow users to actually use color chat with them.

The changes include:
- complete rewrite of the `show_activity`, `show_activity_id` and `show_activity_key` functions - I decided to do this because the code looked very messy, it had repeating lines in all three functions, it was using a "32-loop" instead of `get_players`, etc;
- added a global static bool that will control whether the `show_activity` functions will used colored or default chat messages;
- added a global static integer that will control the `sender` parameter of the `client_print_color` function when used by a `show_activity` function
- added a new stock called `send_activity_message` that will send a message depending on the global bool - if it's true, it will send a message using `client_print_color`, otherwise it will use `client_print`;
- added a new stock called `set_activity_color` that can be used to set whether or not the `show_activity` functions will use colored chat messages, as well as the actual color/sender of them - this functions similarly to `set_hudmessage`, as the parameters from the function stay unchanged until the next call;
- added a few more stocks that prevent usage of repeating code inside of the actual `show_activity` stocks; these stocks include: `get_activity_value` (prepares the `amx_show_activity` cvar for usage, creates a dummy cvar if not found and returns its value as an integer) and `get_activity_prefix` (returns the PLAYER or ADMIN prefix depending on the client's admin status).

With the changes made, here's an example on how to send a colored chat message using the `show_activity` functions:

> set_activity_color(true);
> show_activity_key("ADMIN_SLAP_1", "ADMIN_SLAP_2", name, name2, damage);

Alternatively, the `set_activity_color` function can be used inside `plugin_init` or some other forward in order to make all `show_activity` functions in the code use colored messages.

Another huge problem with these messages is that all of them demand the usage of language keys or format the message in a specific way and they don't allow you to send the message to a specific client or all clients at once. To "fix" this problem, I added a completely new stock called `show_activity_custom` with the following arguments:

> stock show_activity_custom(id, const name[], const fmt[], any:...)

This function combines the power of all other `show_activity` function and allows users to create simple chat messages that obey the `amx_show_activity` cvar. It allows users to send a message to a specific client or all clients at once. It also supports the usage of language files or a simple string. The way it works is that it simply replaces a special placeholder inside the message with the admin's name or completely removes it if the client isn't supposed to see the name. This placeholder is added in amxconst.inc:

> #define ACTIVITY_ADMIN_NAME "$an"

An example of how to use this function to display an activity message to all clients without a language key:

> new szName[32], iValue = 100;
> get_user_name(id, szName, charsmax(szName));
> show_activity_key(0, "ADMIN ^3$an s^1et the cool value to ^4%i", iValue)

OciXCrom commented:
I just got an idea on how to avoid using the huge switch statements in all the functions. I'll update the code tomorrow.

OciXCrom commented:
I updated the code. The functions are now 3-4 times smaller while their functionallity is kept the same.

SmileYzn commented:
Store is_user_admin into a var at stock can_see_admin_name(id)

:v:

OciXCrom commented:
There's no need to do that, it's used only once since it's in a "case".

WPMGPRoSToTeMa commented:
How can I put `$an`/`$px` in the string if I need it? Is there an escape feature for `$`?

OciXCrom commented:
I changed the code a bit. I changed the `set_activity_color` function to `set_activity_rules` and added arguments for changing the $an and $px symbols. Actually a bigger problem would be if player names contain these symbols. Any ideas on how to prevent this? I wasn't able to find a function that gets the buffer without formating it.

I'm not sure if LANG_PLAYER can be even implemented here. If you have an idea how to do it, let me know, otherwise I'll add a @note that LANG_PLAYER isn't supported with these functions.

WPMGPRoSToTeMa commented:
@OciXCrom you can implement it using `SetGlobalTransTarget`.

WPMGPRoSToTeMa commented on plugins/include/amxmisc.inc:
```diff
@@ -390,109 +283,239 @@ stock show_activity_id(idtarget, idadmin, const name[], const fmt[], any:...)
+		replace_all(buffer, len, activity_ph_prefix, prefix);
+	}
+
+	static cansee; cansee = can_see_admin_name(id);
```
Why do you use `static` here?

OciXCrom replied:
Because of my understanding that static should be used in functions that are called each frame. I can see that many people are confused on whether this is true or not - https://forums.alliedmods.net/showthread.php?t=309363

rsKliPPy replied:
Just use `new` everywhere.

OciXCrom commented:
@WPMGPRoSToTeMa @rsKliPPy 

I changed all the `static` declarations with `new` and fixed `LANG_PLAYER`. Let me know if there's something else that can be improved.

OciXCrom commented on plugins/include/amxmisc.inc:
```diff
@@ -337,20 +337,18 @@ stock show_activity_custom(id, const name[], const fmt[], any:...)
-		for(new i; i < pnum; i++)
+		for(new i, player; i < pnum; i++)
 		{
 			player = players[i];
```
Hm? It is `new`.

WPMGPRoSToTeMa replied:
@OciXCrom I mean inside for-loop block.

OciXCrom replied:
@WPMGPRoSToTeMa It's declared inside the for-loop body.

>for(new i, player; i < pnum; i++)

I think you're viewing the old version of the code.

WPMGPRoSToTeMa replied:
@OciXCrom I mean its declaration should be moved to its first assignment.

WPMGPRoSToTeMa replied:
@OciXCrom inside for-loop iteration block. The scope of variable should be as small as possible.
```Pawn
for (new i = 0; i < pnum; i++)
{
	new player = players[i];
```

WPMGPRoSToTeMa replied:
@OciXCrom why not?

OciXCrom replied:
@WPMGPRoSToTeMa Because everyone always says that it's bad for the code. The variable will be created 32 times (in a full server), right now it's declared only once and the scope is the same AFAIK. Even @Arkshine has said this in the forum.

rsKliPPy replied:
It just doesn't matter. Narrowing the scope is better than saving 1 instruction per iteration. A smart compiler (not amxxpc) would optimize that out anyway.
You can also declare it in the first "part" (before the `;`) of the for loop, or just create another scope.

Arkshine replied:
You fool, never quote me!

OciXCrom replied:
@rsKliPPy That's exactly what I did. I declared the variable before the ";" in the loop like I always do, so I don't see a problem. I've always been told that declaring them in the actual loop is bad.

WPMGPRoSToTeMa replied:
@OciXCrom look the difference https://www.diffchecker.com/PUgvCnnc.

WPMGPRoSToTeMa replied:
@OciXCrom did you look into this assembly? Do you have any questions/thoughts?

OciXCrom replied:
@WPMGPRoSToTeMa I can't say that I understand much what that even means. Maybe you can explain? If it's not something extremely important, let's just keep it declared in the for() body for the sake of everyone knowing that that is the better way to do it.

rsKliPPy commented:
It just doesn't matter. Narrowing the scope is better than saving 1 instruction per iteration. A smart compiler (not amxxpc) would optimize that out anyway.
You can also declare it in the first "part" (before the `;`) of the for loop, or just create another scope.

Arkshine commented:
Can you reflect the latest changes in your first post?

OciXCrom commented:
@Arkshine Sorry, I completely forgot about the description there. It's updated now.

OciXCrom commented:
I improved `show_activity_custom` by making it remove the extra whitespace that will appear in the message when the admin's name gets removed.

Before:
```pawn
ADMIN OciXCrom slap Player
ADMIN  slap Player
```

After:
```pawn
ADMIN OciXCrom slap Player
ADMIN slap Player
```
